File: 16. ReverseBits.cpp
// Topic: Reverse bits of an integer for given bit-length
#include <bits/stdc++.h>
using namespace std;

unsigned int reverseBitsN(unsigned int n, int bits)
{
    unsigned int rev = 0;
    for (int i = 0; i < bits; ++i)
    {
        rev = (rev << 1) | (n & 1);
        n >>= 1;
    }
    return rev;
}

int main()
{
    cout << "Reverse 10 in 8 bits: " << reverseBitsN(10, 8) << '\n';
    return 0;
}

// Time Complexity: O(bits)
// Space Complexity: O(1)

File: 16. ReverseBits.py
# Topic: Reverse bits of an integer for given bit-length

def reverse_bits_n(n, bits=32):
    rev = 0
    for _ in range(bits):
        rev = (rev << 1) | (n & 1)
        n >>= 1
    return rev

if __name__ == '__main__':
    print('Reverse 10 in 8 bits:', reverse_bits_n(10,8))

# Time Complexity: O(bits)
# Space Complexity: O(1)
